registry/handlers/images.go

```diff
@@ -11,178 +11,158 @@ import (
+func GetImageManifest(ctx *Context, w http.ResponseWriter, r *http.Request) *HTTPError {
+	ctxu.GetLogger(ctx).Debug("GetImageManifest")
+	tag, dgst := parseTagDigest(ctx, r)
+	manifests := ctx.Repository.Manifests()
 
 	var (
 		sm  *manifest.SignedManifest
 		err error
 	)
 
-	if imh.Tag != "" {
-		sm, err = manifests.GetByTag(imh.Tag)
+	if tag != "" {
+		sm, err = manifests.GetByTag(tag)
 	} else {
-		sm, err = manifests.Get(imh.Digest)
+		sm, err = manifests.Get(dgst)
 	}
 
 	if err != nil {
-		imh.Errors.Push(v2.ErrorCodeManifestUnknown, err)
-		w.WriteHeader(http.StatusNotFound)
-		return
+		return NewHTTPError(v2.ErrorCodeManifestUnknown, err, http.StatusNotFound)
 	}
 
 	// Get the digest, if we don't already have it.
-	if imh.Digest == "" {
-		dgst, err := digestManifest(imh, sm)
+	if dgst == "" {
+		dgst, err = digestManifest(ctx, sm)
 		if err != nil {
-			imh.Errors.Push(v2.ErrorCodeDigestInvalid, err)
-			w.WriteHeader(http.StatusBadRequest)
-			return
+			return NewHTTPError(v2.ErrorCodeDigestInvalid, err, http.StatusBadRequest)
 		}
 
-		imh.Digest = dgst
 	}
 
 	w.Header().Set("Content-Type", "application/json; charset=utf-8")
 	w.Header().Set("Content-Length", fmt.Sprint(len(sm.Raw)))
-	w.Header().Set("Docker-Content-Digest", imh.Digest.String())
+	w.Header().Set("Docker-Content-Digest", dgst.String())
 	w.Write(sm.Raw)
+	return nil
 }
 
 // PutImageManifest validates and stores and image in the registry.
-func (imh *imageManifestHandler) PutImageManifest(w http.ResponseWriter, r *http.Request) {
-	ctxu.GetLogger(imh).Debug("PutImageManifest")
-	manifests := imh.Repository.Manifests()
+func PutImageManifest(ctx *Context, w http.ResponseWriter, r *http.Request) *HTTPError {
+	ctxu.GetLogger(ctx).Debug("PutImageManifest")
+	tag, sentDgst := parseTagDigest(ctx, r)
+	manifests := ctx.Repository.Manifests()
 	dec := json.NewDecoder(r.Body)
 
 	var manifest manifest.SignedManifest
 	if err := dec.Decode(&manifest); err != nil {
-		imh.Errors.Push(v2.ErrorCodeManifestInvalid, err)
-		w.WriteHeader(http.StatusBadRequest)
-		return
+		return NewHTTPError(v2.ErrorCodeManifestInvalid, err, http.StatusBadRequest)
 	}
 
-	dgst, err := digestManifest(imh, &manifest)
+	dgst, err := digestManifest(ctx, &manifest)
 	if err != nil {
-		imh.Errors.Push(v2.ErrorCodeDigestInvalid, err)
-		w.WriteHeader(http.StatusBadRequest)
-		return
+		return NewHTTPError(v2.ErrorCodeDigestInvalid, err, http.StatusBadRequest)
 	}
 
 	// Validate manifest tag or digest matches payload
-	if imh.Tag != "" {
-		if manifest.Tag != imh.Tag {
-			ctxu.GetLogger(imh).Errorf("invalid tag on manifest payload: %q != %q", manifest.Tag, imh.Tag)
-			imh.Errors.Push(v2.ErrorCodeTagInvalid)
-			w.WriteHeader(http.StatusBadRequest)
-			return
+	if tag != "" {
+		if manifest.Tag != tag {
+			ctxu.GetLogger(ctx).Errorf("invalid tag on manifest payload: %q != %q", manifest.Tag, tag)
+			return NewHTTPError(v2.ErrorCodeTagInvalid, nil, http.StatusBadRequest)
```

Comment:
This is why I wanted to add explicit, single, HTTP status codes to the errors defined in v2. I can't serveError on just a v2.ErrorCode because there are multiple possible HTTP Status codes associated with some of the errors.